### PR TITLE
LoadMD needs dimension size for converting

### LINKED ISSUE #21459: LoadMD fails for TOPAZ MD file
### Expected behavior
LoadMD(/SNS/TOPAZ/IPTS-20209/shared/visualization/17pPb_100K/17pPb_100K_combined_symm.nxs) works with Crystallography Convention
### Actual behavior
fails in TransformMD

### PR DESCRIPTION
**Pull request by VickieLynch**
Description of work.
TransposeMD is now working correctly for MD histograms. Also LoadMD was crashing on the first file when loading with Crystallography Convention set.
**To test:**
Set Mantid to Crystallography Convention
```python
Load(Filename='/SNS/TOPAZ/IPTS-20209/shared/visualization/17pPb_100K/17pPb_100K_combined_symm.nxs', OutputWorkspace='17pPb_100K_combined_symm')

data=Load(Filename='CNCS_7860_event.nxs')
data=ConvertUnits(InputWorkspace=data, Target='DeltaE', EMode='Direct', EFixed=3)
data_h=Rebin(InputWorkspace=data, Params='-1,0.01,3', PreserveEvents=False)
md=ConvertToMD(InputWorkspace=data_h, QDimensions='|Q|',  MinValues='0,-1', MaxValues='3,3')
mdh=BinMD(InputWorkspace=md, AlignedDim0='|Q|,0,3,100',AlignedDim1='DeltaE,-1,3,100')
mdh_transf=TransformMD(InputWorkspace=mdh,Scaling='1,-1')
mdh_transf2=TransformMD(InputWorkspace=mdh,Scaling='-1,1')
mdh_transf3=TransformMD(InputWorkspace=mdh,Scaling='-1,-1') 
```

Fixes #21459 

**Release Notes** 

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
17_crystal_inelastic
crystal_inelastic

### CODE REVIEW

**Comment by VickieLynch** [comment]
SliceViewer has a segFault with the tranformation displaying workspace

File: Framework/MDAlgorithms/src/TransformMD.cpp [hunk 1]
```diff
@@ -160,8 +160,24 @@ void TransformMD::exec() {
       signal_t *signals = histo->getSignalArray();
       signal_t *errorsSq = histo->getErrorSquaredArray();
 
-      this->reverse(signals, histo->getNPoints());
-      this->reverse(errorsSq, histo->getNPoints());
+      size_t nPoints = histo->getNPoints();
+      size_t d4 = 1;
+      size_t nd = histo->getNumDims();
+      if (nd > 3) {
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/src/TransformMD.cpp]: These changes mean that I cannot scale my 4th or 5th dimension by -1, even if I want to

Comment (VickieLynch) [reply Framework/MDAlgorithms/src/TransformMD.cpp]: This should be fixed now

File: qt/widgets/common/src/NonOrthogonal.cpp [hunk 2]
```diff
@@ -115,8 +115,15 @@ void doProvideSkewMatrix(Mantid::Kernel::DblMatrix &skewMatrix,
...
     auto const *transform = workspace.getTransformToOriginal();
-    affineMatrix = transform->makeAffineMatrix();
-  } catch (std::runtime_error &) {
+    if (transform)
+      affineMatrix = transform->makeAffineMatrix();
+    else {
+      // Create identity matrix of dimension+1
+      std::size_t nDims = workspace.getNumDims() + 1;
+      Mantid::Kernel::Matrix<Mantid::coord_t> temp(nDims, nDims, true);
+      affineMatrix = temp;
```
Comment (quantumsteve) [inline qt/widgets/common/src/NonOrthogonal.cpp]: Either make line 124 `affineMatrix = std::move(temp);` or make it `affinematrix =  Mantid::Kernel::Matrix<Mantid::coord_t>(nDims, nDims, true);` and remove line 123.

Comment (VickieLynch) [reply qt/widgets/common/src/NonOrthogonal.cpp]: Used second suggestion by Steven

File: Framework/MDAlgorithms/src/TransformMD.cpp [hunk 3]
```diff
@@ -156,13 +156,35 @@ void TransformMD::exec() {
...
-      signal_t *signals = histo->getSignalArray();
-      signal_t *errorsSq = histo->getErrorSquaredArray();
+    // Expect first 3 dimensions to be -1 for changing conventions
+    for (int i = 0; i < static_cast<int>(m_scaling.size()); i++)
+      if (m_scaling[i] < 0) {
+        std::vector<int> axes(m_scaling.size());        // vector with ints.
+        std::iota(std::begin(axes), std::end(axes), 0); // Fill with 0, 1, ...
+        axes[0] = i;
+        axes[i] = 0;
+        // const std::vector<int> axes = { 1, 0 };
```
Comment (SimonHeybrock) [inline Framework/MDAlgorithms/src/TransformMD.cpp]: Stray comment.

File: Framework/MDAlgorithms/src/TransformMD.cpp [hunk 4]
```diff
@@ -156,13 +156,35 @@ void TransformMD::exec() {
...
-    if (m_scaling[0] < 0.0) {
-      signal_t *signals = histo->getSignalArray();
-      signal_t *errorsSq = histo->getErrorSquaredArray();
+    // Expect first 3 dimensions to be -1 for changing conventions
+    for (int i = 0; i < static_cast<int>(m_scaling.size()); i++)
+      if (m_scaling[i] < 0) {
+        std::vector<int> axes(m_scaling.size());        // vector with ints.
+        std::iota(std::begin(axes), std::end(axes), 0); // Fill with 0, 1, ...
+        axes[0] = i;
+        axes[i] = 0;
```
Comment (SimonHeybrock) [inline Framework/MDAlgorithms/src/TransformMD.cpp]: If `i=0` the result of transposing is nothing? Can we start the loop at `1`, or am I missing something?

File: qt/widgets/common/src/NonOrthogonal.cpp [hunk 5]
```diff
@@ -115,8 +115,15 @@ void doProvideSkewMatrix(Mantid::Kernel::DblMatrix &skewMatrix,
...
-  } catch (std::runtime_error &) {
+    if (transform)
+      affineMatrix = transform->makeAffineMatrix();
+    else {
+      // Create identity matrix of dimension+1
+      std::size_t nDims = workspace.getNumDims() + 1;
+      affineMatrix =
+          Mantid::Kernel::Matrix<Mantid::coord_t>(nDims, nDims, true);
+    }
+  } catch (...) {
```
Comment (SimonHeybrock) [inline qt/widgets/common/src/NonOrthogonal.cpp]: New behavior, catching all errors?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
A couple of comments:

File: Framework/MDAlgorithms/src/TransformMD.cpp (hunk 4 above)
Comment (VickieLynch) [reply Framework/MDAlgorithms/src/TransformMD.cpp]: If i=0 and m_scaling[0]<0, the axis is reversed.  This is always true when changing conventions.

File: Framework/MDAlgorithms/src/TransformMD.cpp (hunk 3 above)
Comment (VickieLynch) [reply Framework/MDAlgorithms/src/TransformMD.cpp]: removed

File: Framework/MDAlgorithms/src/TransformMD.cpp (hunk 4 above)
Comment (SimonHeybrock) [reply Framework/MDAlgorithms/src/TransformMD.cpp]: But wouldn't the resulting `axes` be something like `0, 1, 2, 3`, and `TransposeMD` is a no-op?

File: qt/widgets/common/src/NonOrthogonal.cpp (hunk 5 above)
Comment (VickieLynch) [reply qt/widgets/common/src/NonOrthogonal.cpp]: Not sure why I did this.  Changed back to original.

File: Framework/MDAlgorithms/src/TransformMD.cpp (hunk 3 above)
Comment (VickieLynch) [reply Framework/MDAlgorithms/src/TransformMD.cpp]: removed

File: Framework/MDAlgorithms/src/TransformMD.cpp (hunk 4 above)
Comment (VickieLynch) [reply Framework/MDAlgorithms/src/TransformMD.cpp]: transform needed for i=0

File: qt/widgets/common/src/NonOrthogonal.cpp (hunk 5 above)
Comment (VickieLynch) [reply qt/widgets/common/src/NonOrthogonal.cpp]: changed to old error catching

**Review by AndreiSavici — APPROVED** [review]
Looks fine to me. TransformMD does the right thing now for an MD with |Q|, and DeltaE, and it transposes the number of event arrays as well.

File: Framework/MDAlgorithms/src/TransformMD.cpp (hunk 4 above)
Comment (VickieLynch) [reply Framework/MDAlgorithms/src/TransformMD.cpp]: You are right that the transpose does not need to be done for i=0, but the rest with the reverses

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**